Automattic/woocommerce-payments
# Split E2E tests & configure GH actions to run them in parallel

Fixes #3426.

#### Changes proposed in this Pull Request

This PR splits up the existing E2E tests into 3 groups - `WCPay`, `Subscriptions` & `Blocks`. Each group has 2 sub-sections (branches) - `Merchant` & `Shopper`. This allows running E2E tests in parallel to improve test speed & overall time.

Jest configuration is updated to run tests based on the group & branch specified via ENV variables - `E2E_GROUP` & `E2E_BRANCH`. It is also possible to run all tests on a group without specifying the branch variable.

GH Actions workflow is updated to run tests based on the following combinations.

* WCPay - Merchant
* WCPay - Shopper
* Subscriptions - Merchant
* Subscriptions - Shopper
* Blocks - Shopper

Additionally, based on the discussion with @vbelolapotkov, instead of running E2E tests against multiple PHP & WordPress versions, we'll be running tests against the latest WordPress version with PHP 7.4.x. There are no changes to WooCommerce versions being tested.

**E2E Test time comparison**

* Before - 1 hour 10 minutes ([link](https://github.com/Automattic/woocommerce-payments/actions/runs/1520742044))
* After - 23 minutes ([link](https://github.com/Automattic/woocommerce-payments/actions/runs/1520864063))

#### Testing instructions
* On GH Actions, confirm that 25 tests are created under `E2E Tests` and are ran in parallel (the no. of tests ran at a time depends on the number of available GH workers). A new test can be triggered against this branch via `Actions -> E2E Tests -> Run workflow`.
* For testing changes locally, you'll need to specify the env variables along with `npm run test:e2e`. E.g `E2E_GROUP='wcpay' E2E_BRANCH='merchant' npm run test:e2e`

-------------------


**Post merge**
